src/stream/mem_projection.rs

```diff
@@ -0,0 +1,157 @@
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use futures_core::Stream;
+use parquet::{arrow::ProjectionMask, errors::ParquetError};
+use pin_project_lite::pin_project;
+
+use crate::{
+    fs::FileProvider,
+    record::Record,
+    stream::{Entry, ScanStream},
+};
+
+pin_project! {
+    pub struct MemProjectionStream<'projection, R, FP>
+    where
+        R: Record,
+        FP: FileProvider,
+    {
+        stream: Box<ScanStream<'projection, R, FP>>,
+        projection_mask: Arc<ProjectionMask>,
+    }
+}
+
+impl<'projection, R, FP> MemProjectionStream<'projection, R, FP>
+where
+    R: Record,
+    FP: FileProvider + 'projection,
+{
+    pub(crate) fn new(
+        stream: ScanStream<'projection, R, FP>,
+        projection_mask: ProjectionMask,
+    ) -> Self {
+        Self {
+            stream: Box::new(stream),
+            projection_mask: Arc::new(projection_mask),
+        }
+    }
+}
+
+impl<'projection, R, FP> Stream for MemProjectionStream<'projection, R, FP>
+where
+    R: Record,
+    FP: FileProvider + 'projection,
+{
+    type Item = Result<Entry<'projection, R>, ParquetError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut project = self.project();
+
+        return match Pin::new(&mut project.stream).poll_next(cx) {
+            Poll::Ready(Some(Ok(entry))) => Poll::Ready(Some(Ok(Entry::Projection((
+                Box::new(entry),
+                project.projection_mask.clone(),
```

Comment:
Is it possible to avoid `Arc::clone` here?

Reply:
I think it should be OK